rust/src/shards/gui/widgets/text_input.rs

```diff
@@ -0,0 +1,309 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright © 2022 Fragcolor Pte. Ltd. */
+
+use super::TextInput;
+use crate::core::cloneVar;
+use crate::shard::Shard;
+use crate::shards::gui::util;
+use crate::shards::gui::EGUI_UI_SEQ_TYPE;
+use crate::shards::gui::PARENTS_UI_NAME;
+use crate::shards::gui::STRING_VAR_SLICE;
+use crate::shardsc;
+use crate::types::common_type;
+use crate::types::Context;
+use crate::types::ExposedInfo;
+use crate::types::ExposedTypes;
+use crate::types::InstanceData;
+use crate::types::ParamVar;
+use crate::types::Parameters;
+use crate::types::Type;
+use crate::types::Types;
+use crate::types::Var;
+use crate::types::NONE_TYPES;
+use crate::types::STRING_TYPES;
+use std::cmp::Ordering;
+use std::ffi::CStr;
+
+lazy_static! {
+  static ref TEXTINPUT_PARAMETERS: Parameters = vec![(
+    cstr!("Variable"),
+    cstr!("The variable that holds the input value."),
+    STRING_VAR_SLICE,
+  )
+    .into(),];
+}
+
+impl Default for TextInput {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      variable: ParamVar::default(),
+      exposing: Vec::new(),
+      should_expose: false,
+      mutable_text: true,
+    }
+  }
+}
+
+impl AsRef<str> for Var {
```

Comment:
I feel like this might be better as a dedicated function: `as_str` or something.
Also prevents needing the turbofish in http.rs

Reply:
It's required by `egui`. Nothing we can do.

See the definition of the `TextBuffer` trait:
```rust
pub trait TextBuffer: AsRef<str>
```

Reply:
I opened https://github.com/emilk/egui/pull/1824

Reply:
Can you make a type just for the purpose of implementing TextBuffer inside that function?
Something like: 
```rs
enum VarTextBuffer {
 Editable(&mut Var),
 ReadOnly(&Var),
}
```

Since it only really has to live inside that shard's activate function